File: client/lib/common/models/meeting_agenda.dart
enum MeetingAgendaStatus { draft, planned, ongoing, completed }

class MeetingAgenda {
  final int id;
  int number;
  String title;
  String? goals;
  MeetingAgendaStatus status;

  final DateTime redactionDate;
  DateTime? meetingDate;
  String? meetingLocation;

  int? animatorId;
  int? projectId;

  List<int> participantsIds = [];
  List<String> themes = [];

  MeetingAgenda({
    this.id = 0,
    this.number = 0,
    required this.title,
    required this.redactionDate,
    required this.status,
    this.goals = '',
  });

  MeetingAgenda.empty()
    : id = 0,
      number = 0,
      title = '',
      goals = '',
      status = MeetingAgendaStatus.draft,
      redactionDate = DateTime.now();

  MeetingAgenda.fromJson(dynamic data)
    : id = data['id'],
      number = data['number'],
      title = data['title'],
      goals = data['goals'],
      redactionDate = DateTime.parse(data['redactionDate']),
      status = MeetingAgendaStatus.values.firstWhere(
        (e) => e.toString().split('.').last == data['status'],
      ),
      meetingDate =
          data['meetingDate'] != null
              ? DateTime.parse(data['meetingDate'])
              : null,
      meetingLocation = data['meetingLocation'],
      animatorId = data['animatorId'],
      participantsIds = List<int>.from(data['participantsIds']),
      themes = List<String>.from(data['themes']),
      projectId = data['projectId'];

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'number': number,
      'title': title,
      'redactionDate': redactionDate.toIso8601String(),
      'status': status.toString().split('.').last,

      if (goals != null) 'goals': goals,
      if (meetingDate != null) 'meetingDate': meetingDate!.toIso8601String(),
      if (meetingLocation != null) 'meetingLocation': meetingLocation,
      if (animatorId != null) 'animatorId': animatorId,
      if (participantsIds.isNotEmpty) 'participantsIds': participantsIds,
      if (themes.isNotEmpty) 'themes': themes,
      if (projectId != null) 'projectId': projectId,
    };
  }
}

File: server/flaskr/models/data/meeting_agenda.py
from dataclasses import dataclass
from enum import Enum
from datetime import datetime


class MeetingAgendaStatus(Enum):
    draft = 'draft'
    planned = 'planned'
    completed = 'completed'

@dataclass(slots=True)
class MeetingAgenda:
    id: str
    number: int
    title: str
    goals: str
    status: MeetingAgendaStatus

    redactionDate: datetime
    meetingDate: datetime
    meetingLocation: str

    animatorId: str
    projectId: str

    participantsIds: list[str]
    themes: list[str]

    def to_dict(self):
        return {
            "id": self.id,
            "number": self.number,
            "title": self.title,
            "goals": self.goals,
            "status": self.status,
            "redactionDate": self.redactionDate.isoformat(),
            "meetingDate": self.meetingDate.isoformat() if self.meetingDate else None,
            "meetingLocation": self.meetingLocation,
            "animatorId": self.animatorId,
            "participantsIds": self.participantsIds or [],
            "themes": self.themes or [],
            "projectId": self.projectId,
        }
